pkg/controller/controller.go

```diff
@@ -997,6 +1070,150 @@ func (p *csiProvisioner) Delete(volume *v1.PersistentVolume) error {
+	}
+}
+
+// processNextClaimWorkItem processes items from claimQueue
+func (p *csiClaimController) processNextClaimWorkItem() bool {
+	obj, shutdown := p.claimQueue.Get()
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer p.claimQueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			p.claimQueue.Forget(obj)
+			return fmt.Errorf("expected string in workqueue but got %#v", obj)
+		}
+
+		if err := p.syncClaimHandler(key); err != nil {
+			klog.Warningf("Retrying syncing claim %q after %v failures", key, p.claimQueue.NumRequeues(obj))
+			p.claimQueue.AddRateLimited(obj)
+		}
+
+		p.claimQueue.Forget(obj)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// enqueueClaimUpadate takes an obj and converts it into UID that is then put onto claim work queue.
+func (p *csiClaimController) enqueueClaimUpadate(obj interface{}) {
+	new, ok := obj.(*v1.PersistentVolumeClaim)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expected claim but got %+v", new))
+		return
+	}
+
+	// Timestamp didn't appear
+	if new.DeletionTimestamp == nil {
+		return
+	}
+
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	p.claimQueue.Add(key)
+}
+
+// syncClaimHandler gets the claim from informer's cache then calls syncClaim
+func (p *csiClaimController) syncClaimHandler(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	claimObj, err := p.claimLister.PersistentVolumeClaims(namespace).Get(name)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("Item '%s' in work queue no longer exists", key))
+			return nil
+		}
+
+		return err
+	}
+
+	return p.syncClaim(claimObj)
+}
+
+func (p *csiClaimController) syncClaim(obj interface{}) error {
+	claim, ok := obj.(*v1.PersistentVolumeClaim)
+	if !ok {
+		return fmt.Errorf("expected claim but got %+v", obj)
+	}
+
+	if !checkFinalizer(claim, pvcCloneFinalizer) {
+		return nil
+	}
+
+	pvcList, err := p.client.CoreV1().PersistentVolumeClaims("").List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Check for pvc state with DataSource pointing to claim
+	for _, pvc := range pvcList.Items {
+		if pvc.Spec.DataSource == nil {
+			continue
+		}
+
+		if pvc.Spec.DataSource.Kind == pvcKind &&
+			pvc.Spec.DataSource.Name == claim.Name &&
+			pvc.Status.Phase == v1.ClaimPending {
+			return apierrs.NewBadRequest(fmt.Sprintf("PVC '%s' is in pending state, cloning in progress", pvc.Name))
+		}
+	}
+
+	// Remove clone finalizer
+	// need to get the pvc again because the delete has updated the object with a deletion timestamp
+	claim, err = p.client.CoreV1().PersistentVolumeClaims(claim.Namespace).Get(claim.Name, metav1.GetOptions{})
```

Comment:
I don't understand why extra `Get()` is necessary. `claim` already has `DeletionTimestamp`. Yes, it could have changed since the informer got update event from API server, but it could have changed also since you get it from API server here and use the information few lines below.